### ISSUE TITLE
Need help in extracting logits (token + probabilities)! 

### ISSUE THREAD

**Issue by AkshathRaghav** [ggerganov/llama.cpp#6285]
Hi guys, I'm experimenting with llama.cpp for a project, and I'd like to get all the logits (tokens + probabilities) from llama.cpp. 

I tried the save-logits and logits-all flags from main.cpp, but it returns the tokens and token ids directly into the log file. My goal is to get the logits to understand sampling and how it works. (I have a high level understanding, but want to actually get my hands-dirty). 

Could someone help me out on this? I'd appreciate any direction in getting these values out. Thank you! 

Also, is there a way to do this without making changes into the code of llama.cpp? That would be even better and easier.

**Comment by ngxson** [comment]
You can base on the `simple.cpp` example, which extract the logits and use greedy method to sort and sample next token:

https://github.com/ggerganov/llama.cpp/blob/a0e584defd8c16e7a51ab895f595df0448d710d0/examples/simple/simple.cpp#L128

To read out the list of tokens from `std::vector<llama_token_data> candidates;`

```cpp
for (auto & candidate : candidates) {
  LOG_TEE("%d (logit=%f) --> %s", candidate.id, candidate.logit, llama_token_to_piece(ctx, candidate.id).c_str());
}
```
